src/Build/Instance/TaskRegistry.cs

```diff
@@ -660,7 +693,45 @@ private void RegisterTask(string taskName, AssemblyLoadInfo assemblyLoadInfo, st
                 _taskRegistrations[taskIdentity] = registeredTaskEntries;
             }
 
-            registeredTaskEntries.Add(new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord));
+            RegisteredTaskRecord newRecord = new RegisteredTaskRecord(taskName, assemblyLoadInfo, taskFactory, taskFactoryParameters, inlineTaskRecord);
+
+            if (overrideTask)
+            {
+                // Key the dictionary based on UNqualified task names
+                // This is to support partial matches on tasks like Foo.Bar and Baz.Bar
+                string unqualifiedTaskName = taskName;
+
+                if (unqualifiedTaskName.Contains('.'))
+                {
+                    unqualifiedTaskName = taskName.Split('.').Last();
+                }
+
+                // Is the task already registered?
+                if (overriddenTasks.TryGetValue(unqualifiedTaskName, out List<RegisteredTaskRecord> recs))
+                {
+                    // check every registration that exists in the list.
+                    foreach (RegisteredTaskRecord rec in recs)
+                    {
+                        // Does the same registration already exist? (same exact name)
+                        if (rec.RegisteredName.Equals(taskIdentity.Name, StringComparison.OrdinalIgnoreCase))
+                        {
+                            loggingService.LogError(context, null, new BuildEventFileInfo(projectUsingTaskInXml.OverrideLocation), "DuplicateOverrideUsingTaskElement", taskName);
+                            break;
+                        }
+                    }
+                    recs.Add(newRecord);
+                }
+                else
+                {
+                    // Create a dictionary containing the unqualified name (for quick lookups when the task is called).
+                    // Place the new record using the potentially fully qualified name to account for partial matches.
+                    overriddenTasks.Add(unqualifiedTaskName, new List<RegisteredTaskRecord>());
+                    overriddenTasks[unqualifiedTaskName].Add(newRecord);
```

Comment:
```suggestion
                    // New record's name may be fully qualified. Use it anyway to account for partial matches.
                    List<RegisteredTaskRecord> unqualifiedTaskNameMatches = new();
                    unqualifiedTaskNameMatches.Add(newRecord);
                    overriddenTasks.Add(unqualifiedTaskName, unqualifiedTaskNameMatches);
```
?

Saves a dictionary access, too.